crates/matrix-sdk/src/latest_events/mod.rs

```diff
@@ -0,0 +1,861 @@
+        let (room_registration_sender, room_registration_receiver) = mpsc::channel(32);
+        let (latest_event_queue_sender, latest_event_queue_receiver) = mpsc::unbounded_channel();
+
+        let registered_rooms = Arc::new(RegisteredRooms::new(room_registration_sender));
+
+        // The task listening to the event cache and the send queue updates.
+        let listen_task_handle = spawn(listen_to_event_cache_and_send_queue_updates_task(
+            registered_rooms.clone(),
+            room_registration_receiver,
+            event_cache,
+            send_queue,
+            latest_event_queue_sender,
+        ))
+        .abort_on_drop();
+
+        // The task computing the new latest events.
+        let computation_task_handle = spawn(compute_latest_events_task(
+            registered_rooms.clone(),
+            latest_event_queue_receiver,
+        ))
+        .abort_on_drop();
+
+        Self {
+            state: Arc::new(LatestEventsState {
+                registered_rooms,
+                _listen_task_handle: listen_task_handle,
+                _computation_task_handle: computation_task_handle,
+            }),
+        }
+    }
+
+    /// Start listening to updates (if not already) for a particular room, and
+    /// fetch the [`LatestEvent`] for this room.
+    ///
+    /// It returns `None` if the room doesn't exist.
+    ///
+    /// Note that a read lock guard is returned if the [`LatestEvent`] exists.
+    /// One doesn't want to keep this lock alive for too long, otherwise the
+    /// whole [`LatestEvents`] API is blocked. Use [`LatestEvent::get`] or
+    /// [`LatestEvent::subscribe`] as quick as possible, and then drop this
+    /// lock guard.
+    pub async fn for_room(&self, room_id: &RoomId) -> Option<RwLockReadGuard<'_, LatestEvent>> {
+        self.state.registered_rooms.for_room(room_id).await
+    }
+
+    /// Start listening to updates (if not already) for a particular room, and
+    /// fetch the [`LatestEvent`] for a particular thread in this room.
+    ///
+    /// It returns `None` if the room or the thread doesn't exist.
+    ///
+    /// Note that a read lock guard is returned if the [`LatestEvent`] exists.
+    /// One doesn't want to keep this lock alive for too long, otherwise the
+    /// whole [`LatestEvents`] API is blocked. Use [`LatestEvent::get`] or
+    /// [`LatestEvent::subscribe`] as quick as possible, and then drop this
+    /// lock guard.
+    pub async fn for_thread(
+        &self,
+        room_id: &RoomId,
+        thread_id: &EventId,
+    ) -> Option<RwLockReadGuard<'_, LatestEvent>> {
+        self.state.registered_rooms.for_thread(room_id, thread_id).await
+    }
+
+    /// Forget a room.
+    ///
+    /// It means that [`LatestEvents`] will stop listening to updates for the
+    /// `LatestEvent`s of the room and all its threads.
+    ///
+    /// If [`LatestEvents`] is not listening for `room_id`, nothing happens.
+    pub async fn forget_room(&self, room_id: &RoomId) {
```

Comment:
Can't help but wonder, is there any version of this in which the user doesn't need to care about deregestering from subscribers manually? If I get a `LatestEvent` and subscribe to its values, can dropping that take care of the forgetting  maybe?